trace-telemetry-exporter/src/main.rs

```diff
@@ -0,0 +1,228 @@
+
+#[derive(Debug, Parser)]
+#[clap(author, version, about)]
+struct Cli {
+    #[clap(long)]
+    broker: String,
+
+    #[clap(long)]
+    username: Option<String>,
+
+    #[clap(long)]
+    password: Option<String>,
+
+    #[clap(long = "group")]
+    consumer_group: String,
+
+    #[clap(long)]
+    trace_topic: String,
+
+    #[clap(long, env, default_value = "127.0.0.1:9091")]
+    metrics_address: SocketAddr,
+
+    #[clap(long, default_value_t = 5)]
+    message_rate_interval: u64,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt::init();
+
+    let args = Cli::parse();
+
+    let consumer: StreamConsumer = supermusr_common::generate_kafka_client_config(
+        &args.broker,
+        &args.username,
+        &args.password,
+    )
+    .set("group.id", &args.consumer_group)
+    .set("enable.partition.eof", "false")
+    .set("session.timeout.ms", "6000")
+    .set("enable.auto.commit", "false")
+    .create()?;
+
+    consumer.subscribe(&[&args.trace_topic])?;
+
+    let builder = PrometheusBuilder::new();
+    builder
+        .with_http_listener(args.metrics_address)
+        .install()
+        .expect("prometheus metrics exporter should be setup");
+
+    metrics::describe_counter!(
+        METRIC_MSG_COUNT,
+        metrics::Unit::Count,
+        "Number of messages received"
+    );
+
+    metrics::describe_counter!(
+        METRIC_MSG_RATE,
+        metrics::Unit::CountPerSecond,
+        "Rate of messages received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_LAST_MSG_TIMESTAMP,
+        metrics::Unit::Nanoseconds,
+        "Timestamp of last message received (ns)"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_LAST_MSG_FRAME_NUMBER,
+        metrics::Unit::Count,
+        "Frame number of last message received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_CHANNEL_COUNT,
+        metrics::Unit::Count,
+        "Number of channels in last message received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_SAMPLE_COUNT,
+        metrics::Unit::Count,
+        "Number of samples in last message received"
+    );
+
+    let recent_msg_counts: MessageCounts = Arc::new(Mutex::new(HashMap::new()));
+
+    task::spawn(update_message_rate(
+        Arc::clone(&recent_msg_counts),
+        args.message_rate_interval,
+    ));
+
+    poll_kafka_msg(consumer).await;
+
+    Ok(())
+}
+
+fn get_digitiser_label(digitiser_id: u8) -> (String, String) {
+    ("digitiser_id".to_string(), format!("{}", digitiser_id))
+}
+
+async fn update_message_rate(recent_msg_counts: MessageCounts, recent_message_lifetime: u64) {
+    loop {
+        // Wait a set period of time before calculating average.
+        sleep(Duration::from_secs(recent_message_lifetime)).await;
+        let mut recent_msg_counts = recent_msg_counts.lock().unwrap();
+        // Calculate and record message rate for each digitiser.
+        recent_msg_counts.iter().for_each(|recent_msg_count| {
```

Comment:
You should be able to expand this in the argument list.
```suggestion
        recent_msg_counts.iter().for_each(|(digitiser_id, recent_msg_count)| {
```